### PR TITLE
[FLINK-4149] Fix Serialization of NFA in AbstractKeyedCEPPatternOperator

### PR DESCRIPTION
**Pull request by aljoscha**
NFA is Serializable and has readObject()/writeObject() methods. In
AbstractKeyedCEPPatternOperator a KryoSerializer was used as the
TypeSerializer for the ValueState that holds NFA instances. Kryo does
not call readObject()/writeObject() therefore the state of the NFA was
invalid after deserialization.

This change adds a new TypeSerializer for NFA that uses
Java Serialization. In the long run it will be better to get rid of the
readObject()/writeObject() methods and instead efficiently serialize
using a specialized TypeSerializer.

This also adds a test that verifies that checkpoint/restore work as expected.

### CODE REVIEW

**Comment by aljoscha** [comment]
R: @tillrohrmann for review

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 1]
```diff
@@ -19,12 +19,16 @@
 package org.apache.flink.cep.nfa;
 
 import com.google.common.collect.LinkedHashMultimap;
+import org.apache.commons.io.output.ByteArrayOutputStream;
```
Comment (tillrohrmann) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Why do we use once the `org.apache.commons.io.output.BAOS` and then the `java.io.BAIS`?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 2]
```diff
@@ -454,4 +458,126 @@ static String generateStateName(final String name, final int index) {
...
+
+				oos.writeObject(record);
+
+				oos.close();
+				baos.close();
+
+				byte[] data = baos.toByteArray();
+
+				target.writeInt(data.length);
+				target.write(data);
```
Comment (tillrohrmann) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Can't we use the `DataOutputViewStream`?

```
ObjectOutputStream oos = new ObjectOutputStream(new DataOutputViewStream(target));
oos.writeObject(record);
oos.close();
```

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 3]
```diff
@@ -454,4 +458,126 @@ static String generateStateName(final String name, final int index) {
...
+				byte[] data = new byte[size];
+
+				source.readFully(data);
+
+				ByteArrayInputStream bais = new ByteArrayInputStream(data);
+				ObjectInputStream ois = new ObjectInputStream(bais);
+
+				@SuppressWarnings("unchecked")
+				NFA<T> copy = (NFA<T>) ois.readObject();
+				return copy;
```
Comment (tillrohrmann) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Same here with the `DataInputViewStream`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 4]
```diff
@@ -454,4 +458,126 @@ static String generateStateName(final String name, final int index) {
...
+		@Override
+		public void copy(DataInputView source, DataOutputView target) throws IOException {
+			int size = source.readInt();
+			target.writeInt(size);
+			target.write(source, size);
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			return obj instanceof Serializer;
```
Comment (tillrohrmann) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: In order to make the equals relation symmetric, we should `return obj instanceof Serializer && ((Serializer) obj).canEqual(this);`.

**Comment by tillrohrmann** [comment]
Changes look good to me. Thanks for fixing the CEP issues @aljoscha :-). I had some minor comments concerning the NFA serializer.

Before merging it would be good to rebase on the latest master and rerun Travis, because many builds failed due to a problem with the Kinesis connector. Thus, there might be some problems shadowed by the failing Kinesis tests (not really expecting this).

**Comment by aljoscha** [comment]
Thanks for the review @tillrohrmann! The `commons.io` was just an IDE mess-up. The `equals` I will fix.

**Comment by shashank734** [comment]
Again getting this error. After changing state backend from Rocksdb it's working fine.

```
04/12/2017 10:05:04	Job execution switched to status FAILING.
java.lang.RuntimeException: Could not deserialize NFA.
	at org.apache.flink.cep.nfa.NFA$Serializer.deserialize(NFA.java:538)
	at org.apache.flink.cep.nfa.NFA$Serializer.deserialize(NFA.java:469)
	at org.apache.flink.contrib.streaming.state.RocksDBValueState.value(RocksDBValueState.java:81)
	at org.apache.flink.cep.operator.AbstractKeyedCEPPatternOperator.getNFA(AbstractKeyedCEPPatternOperator.java:124)
	at org.apache.flink.cep.operator.AbstractCEPBasePatternOperator.processElement(AbstractCEPBasePatternOperator.java:72)
	at org.apache.flink.cep.operator.AbstractKeyedCEPPatternOperator.processElement(AbstractKeyedCEPPatternOperator.java:162)
	at org.apache.flink.streaming.runtime.io.StreamInputProcessor.processInput(StreamInputProcessor.java:185)
	at org.apache.flink.streaming.runtime.tasks.OneInputStreamTask.run(OneInputStreamTask.java:63)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:272)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:655)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: co.ronak.nto.Job$$anon$18$$anon$21$$anon$3
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at java.io.ObjectInputStream.resolveClass(ObjectInputStream.java:626)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1774)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
	at java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:501)
	at org.apache.flink.api.scala.typeutils.TraversableSerializer.readObject(TraversableSerializer.scala:53)
	at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1017)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1900)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1707)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1345)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
	at java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:501)
	at org.apache.flink.cep.NonDuplicatingTypeSerializer.readObject(NonDuplicatingTypeSerializer.java:190)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1017)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1900)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
	at java.io.ObjectInputStream.defaultReadObject(ObjectInputStream.java:501)
	at org.apache.flink.cep.nfa.NFA.readObject(NFA.java:394)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1017)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1900)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
	at org.apache.flink.cep.nfa.NFA$Serializer.deserialize(NFA.java:535)
	... 10 more
```

**Comment by aljoscha** [comment]
Thanks @shashank734 I created another issue since this is a different (but somewhat related) problem: https://issues.apache.org/jira/browse/FLINK-6318

**Pull request closed without merging**